pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -96,192 +143,309 @@ func shouldRestartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) bool {
-	leaderPhase, err := encryptionKeyRotationPhaseFromPeriodic(leader)
+// encryptionKeyRotationRestartNodes will restart the leader to ensure that the current encryption key rotation status
+// is persisted to the datastore. With a single node control plane, the apiserver container needs to be restarted,
+// however with an HA control plane, the leader must first be restarted then all the followers, as the followers will
+// update their status when reading from the datastore on startup of the rke2/k3s services.
+func (p *Planner) encryptionKeyRotationRestartNodes(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
+	if rke2.GetRuntime(cp.Spec.KubernetesVersion) == rke2.RuntimeRKE2 &&
+		len(collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)) == 1 {
+		nodePlan := plan.NodePlan{
+			Files: []plan.File{
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationRestartAPIServerScript)),
+					Path:    encryptionKeyRotationRestartAPIServerScriptPath,
+				},
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationGetAPIServerStatusScript)),
+					Path:    encryptionKeyRotationGetAPIServerStatusScriptPath,
+				},
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationWaitForSecretsEncryptStatusScript)),
+					Path:    encryptionKeyRotationWaitForSecretsEncryptStatusScriptPath,
+				},
+			},
+			Instructions: []plan.OneTimeInstruction{
+				p.encryptionKeyRotationRestartApiserver(cp),
+				encryptionKeyRotationWaitForSecretsEncryptStatus(cp, ""),
+				encryptionKeyRotationSecretsEncryptStatusOneTimeInstruction(cp, ""),
+			},
+			PeriodicInstructions: []plan.PeriodicInstruction{
+				p.encryptionKeyRotationApiserverStatus(cp),
+			},
+		}
+		err := assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
+		if err != nil {
+			return p.enqueueIfErrWaiting(cp, err)
+		}
+
+		if _, ok := cp.Annotations[encryptionKeyRotationApiserverIDAnnotation]; !ok {
+			apiserverID := strings.TrimSuffix(string(leader.Plan.Output[encryptionKeyRotationRestartApiserverCommand]), "\n")
+			if len(apiserverID) != 13 { // not a valid container id
+				return p.enqueueAndSkip(cp)
+			}
+			cp.Annotations[encryptionKeyRotationApiserverIDAnnotation] = apiserverID
+			cp, err = p.rkeControlPlanes.Update(cp)
+			if err != nil {
+				return err
+			}
+			return p.enqueueAndSkip(cp)
+		}
+
+		apiserverStatus := strings.TrimSuffix(string(leader.Plan.PeriodicOutput[encryptionKeyRotationApiserverStatusCommand].Stdout), "\n")
+		if apiserverStatus == "" {
+			return p.enqueueAndSkip(cp)
+		}
+		if cp.Annotations[encryptionKeyRotationApiserverIDAnnotation] == apiserverStatus {
+			return p.enqueueAndSkip(cp)
+		}
+		if len(apiserverStatus) != 13 { // not a valid container id
+			return p.enqueueAndSkip(cp)
+		}
+	} else {
+		nodePlan := plan.NodePlan{
```

Comment:
this seems to be used if we are not RKE2 or if we have an RKE2 cluster with more than 1 suitable controlplane node?